net/tshttpproxy/tshttpproxy_synology.go

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build linux
+// +build linux
+
+package tshttpproxy
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"tailscale.com/util/lineread"
+	"tailscale.com/version/distro"
+)
+
+func init() {
+	sysProxyFromEnv = linuxSysProxyFromEnv
+}
+
+var (
+	distroOnce         sync.Once
+	distroProxyFromEnv func(*http.Request) (*url.URL, error)
+)
+
+func linuxSysProxyFromEnv(req *http.Request) (*url.URL, error) {
+	distroOnce.Do(func() {
+		if distro.Get() == distro.Synology {
+			distroProxyFromEnv = synologyProxyFromConfigCached
+			return
+		}
+	})
+	if distroProxyFromEnv != nil {
+		return distroProxyFromEnv(req)
+	}
+	return nil, nil
+}
+
+var cache struct {
+	sync.Mutex
+	proxy   *url.URL
+	updated time.Time
+}
+
+func synologyProxyFromConfigCached(req *http.Request) (*url.URL, error) {
+	cache.Lock()
+	defer cache.Unlock()
+
+	if !cache.updated.IsZero() && time.Since(cache.updated) < time.Minute {
+		return cache.proxy, nil
+	}
+
+	cache.updated = time.Now()
+
+	val, err := synologyProxyFromConfig(req)
+	cache.proxy = val
+	return val, err
+}
+
+func synologyProxyFromConfig(req *http.Request) (*url.URL, error) {
+	if req.URL == nil {
+		return nil, nil
+	}
+
+	r, err := openSynologyProxyConf()
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	defer r.Close()
+
+	return parseSynologyConfig(r)
+}
+
+var openSynologyProxyConf = func() (io.ReadCloser, error) {
```

Comment:
This looks very weird without a comment saying it's a `// variable for testing` or something